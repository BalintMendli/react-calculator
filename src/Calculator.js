import React from 'react';
import PropTypes from 'prop-types';

const Calculator = ({
  display,
  handleClear,
  handleEval,
  handleNum,
  handleOp,
  handlePoint,
}) => (
  <div id="calc-div">
    <div id="display">{display}</div>
    <button type="button" id="clear" className="buttons" onClick={handleClear}>
      AC
    </button>
    <button type="button" id="divide" className="buttons" onClick={handleOp}>
      /
    </button>
    <button type="button" id="multiply" className="buttons" onClick={handleOp}>
      *
    </button>
    <button type="button" id="seven" className="buttons" onClick={handleNum}>
      7
    </button>
    <button type="button" id="eight" className="buttons" onClick={handleNum}>
      8
    </button>
    <button type="button" id="nine" className="buttons" onClick={handleNum}>
      9
    </button>
    <button type="button" id="subtract" className="buttons" onClick={handleOp}>
      -
    </button>
    <button type="button" id="four" className="buttons" onClick={handleNum}>
      4
    </button>
    <button type="button" id="five" className="buttons" onClick={handleNum}>
      5
    </button>
    <button type="button" id="six" className="buttons" onClick={handleNum}>
      6
    </button>
    <button type="button" id="add" className="buttons" onClick={handleOp}>
      +
    </button>
    <button type="button" id="one" className="buttons" onClick={handleNum}>
      1
    </button>
    <button type="button" id="two" className="buttons" onClick={handleNum}>
      2
    </button>
    <button type="button" id="three" className="buttons" onClick={handleNum}>
      3
    </button>
    <button type="button" id="equals" className="buttons" onClick={handleEval}>
      =
    </button>
    <button type="button" id="zero" className="buttons" onClick={handleNum}>
      0
    </button>
    <button
      type="button"
      id="decimal"
      className="buttons"
      onClick={handlePoint}
    >
      .
    </button>
  </div>
);

Calculator.propTypes = {
  display: PropTypes.string.isRequired,
  handleClear: PropTypes.func.isRequired,
  handleEval: PropTypes.func.isRequired,
  handleNum: PropTypes.func.isRequired,
  handleOp: PropTypes.func.isRequired,
  handlePoint: PropTypes.func.isRequired,
};

export default Calculator;
