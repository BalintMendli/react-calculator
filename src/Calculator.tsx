import React from 'react';

interface CalculatorProps {
  display: string;
  handleClear: React.MouseEventHandler<HTMLButtonElement>;
  handleEval: React.MouseEventHandler<HTMLButtonElement>;
  handleNum: React.MouseEventHandler<HTMLButtonElement>;
  handleOp: React.MouseEventHandler<HTMLButtonElement>;
  handlePoint: React.MouseEventHandler<HTMLButtonElement>;
}

const Calculator = ({
  display,
  handleClear,
  handleEval,
  handleNum,
  handleOp,
  handlePoint,
}: CalculatorProps): JSX.Element => (
  <div id="calc-div">
    <div id="display">{display}</div>
    <button type="button" id="clear" className="buttons" onClick={handleClear}>
      AC
    </button>
    <button type="button" id="divide" className="buttons" onClick={handleOp}>
      /
    </button>
    <button type="button" id="multiply" className="buttons" onClick={handleOp}>
      *
    </button>
    <button type="button" id="seven" className="buttons" onClick={handleNum}>
      7
    </button>
    <button type="button" id="eight" className="buttons" onClick={handleNum}>
      8
    </button>
    <button type="button" id="nine" className="buttons" onClick={handleNum}>
      9
    </button>
    <button type="button" id="subtract" className="buttons" onClick={handleOp}>
      -
    </button>
    <button type="button" id="four" className="buttons" onClick={handleNum}>
      4
    </button>
    <button type="button" id="five" className="buttons" onClick={handleNum}>
      5
    </button>
    <button type="button" id="six" className="buttons" onClick={handleNum}>
      6
    </button>
    <button type="button" id="add" className="buttons" onClick={handleOp}>
      +
    </button>
    <button type="button" id="one" className="buttons" onClick={handleNum}>
      1
    </button>
    <button type="button" id="two" className="buttons" onClick={handleNum}>
      2
    </button>
    <button type="button" id="three" className="buttons" onClick={handleNum}>
      3
    </button>
    <button type="button" id="equals" className="buttons" onClick={handleEval}>
      =
    </button>
    <button type="button" id="zero" className="buttons" onClick={handleNum}>
      0
    </button>
    <button
      type="button"
      id="decimal"
      className="buttons"
      onClick={handlePoint}
    >
      .
    </button>
  </div>
);

export default Calculator;
