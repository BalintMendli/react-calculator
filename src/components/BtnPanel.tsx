import React from 'react';
import styled from 'styled-components/macro';
import buttons, { Label } from 'constants/buttons';
import { getColor } from 'constants/colors';
import { rows, columns } from 'constants/dimensions';
import { getDimStyle } from 'logic/utils';
import Button from './Button';

interface BtnPanelProps {
  handleClick: (label: Label) => void;
}

interface StyledProps {
  width: number;
  height: number;
}

const StyledBtnPanel = styled.div<StyledProps>`
  background-color: #0b0c11;
  flex: 1;
  display: grid;
  grid-template-columns: ${(props): string => getDimStyle(props.width)};
  grid-template-rows: ${(props): string => getDimStyle(props.height)};
  box-shadow: 1px 2px 10px rgba(0, 0, 0, 0.5);
  border-radius: 5px;
`;

const BtnPanel = ({ handleClick }: BtnPanelProps): JSX.Element => (
  <StyledBtnPanel id="btn-panel" width={columns} height={rows}>
    {buttons.map(btn => (
      <Button
        key={btn.label}
        label={btn.label}
        id={btn.id}
        pos={btn.position}
        color={getColor(btn)}
        handleClick={handleClick}
      />
    ))}
  </StyledBtnPanel>
);

export default BtnPanel;
