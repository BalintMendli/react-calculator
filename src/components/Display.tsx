import React from 'react';
import styled from 'styled-components/macro';
import { transformForDisplay, transformError } from 'logic/utils';

interface DisplayProps {
  result: string;
  expression: string;
  error: string;
}

const StyledDisplay = styled.div`
  height: 80px;
  color: #fff;
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  justify-content: flex-start;
  padding-bottom: 3px;
  font-size: 28px;
  line-height: 1.05;
`;

const StyledError = styled.div`
  height: 18px;
  max-width: 420px;
  padding: 2px 5px;
  margin-top: 5px;
  background-color: #a5243d;
  font-size: 12px;
  border-radius: 2px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;

  :empty {
    visibility: hidden;
  }
`;

const StyledExp = styled.div`
  height: 22px;
  font-size: 16px;
  display: flex;
  align-items: center;
  transition: all 150ms ease-in;
  transform-origin: top right;

  :empty {
    transform: translate(-15px, 25px) scale(1.5);
    opacity: 0;
  }
`;

const Display = ({ result, expression, error }: DisplayProps): JSX.Element => (
  <StyledDisplay id="display">
    <StyledError id="error">
      {error ? transformError(error, result) : ''}
    </StyledError>
    <StyledExp id="expression">
      {expression ? `${transformForDisplay(expression)} =` : ''}
    </StyledExp>
    <div id="result">{transformForDisplay(result)}</div>
  </StyledDisplay>
);

export default Display;
