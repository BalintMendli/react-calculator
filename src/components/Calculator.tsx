import React from 'react';
import styled from 'styled-components/macro';
import { Label } from 'constants/buttons';
import { heightPx, widthPx } from 'constants/dimensions';
import Display from './Display';
import BtnPanel from './BtnPanel';

interface CalculatorProps {
  result: string;
  expression: string;
  error: string;
  handleClick: (label: Label) => void;
}

interface StyledProps {
  widthPx: number;
  heightPx: number;
}

const StyledCalculator = styled.div<StyledProps>`
  height: ${(props): string => props.heightPx.toString()}px;
  width: ${(props): string => props.widthPx.toString()}px;
  background-color: #0b0c11;
  padding: 5px;
  display: flex;
  flex-direction: column;
  box-shadow: 1px 2px 10px rgba(0, 0, 0, 0.5);
  border-radius: 5px;
`;

const Calculator = ({
  result,
  expression,
  error,
  handleClick,
}: CalculatorProps): JSX.Element => (
  <StyledCalculator id="calculator" widthPx={widthPx} heightPx={heightPx}>
    <Display result={result} expression={expression} error={error} />
    <BtnPanel handleClick={handleClick} />
  </StyledCalculator>
);

export default Calculator;
