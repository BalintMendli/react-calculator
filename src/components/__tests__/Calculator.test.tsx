import React from 'react';
import Calculator from 'components/Calculator';
import Display from 'components/Display';
import BtnPanel from 'components/BtnPanel';
import { shallow, ShallowWrapper } from 'enzyme';

describe('Calculator component', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Calculator
        error="test error"
        expression="test expression"
        result="test result"
        handleClick={jest.fn()}
      />,
    );
  });

  it('renders correctly', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('renders calculator div', () => {
    expect(wrapper.find('#calculator').length).toEqual(1);
  });

  it('renders Display', () => {
    expect(wrapper.find(Display).length).toEqual(1);
  });

  it('renders BtnPanel', () => {
    expect(wrapper.find(BtnPanel).length).toEqual(1);
  });
});
