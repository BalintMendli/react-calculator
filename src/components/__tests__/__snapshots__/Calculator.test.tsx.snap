// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component renders correctly 1`] = `
"<Calculator__StyledCalculator id=\\"calculator\\" widthPx={450} heightPx={350}>
  <Display result=\\"test result\\" expression=\\"test expression\\" error=\\"test error\\" />
  <BtnPanel handleClick={[Function: mockConstructor]} />
</Calculator__StyledCalculator>"
`;
